### PR TITLE
fix: `ListUsers` handling of cycles in exclusion handler

### PR DESCRIPTION
**Pull request by willvedd**
## Description
As noted in a [recent CVE advisory](https://github.com/openfga/openfga/security/advisories/GHSA-8cph-m685-6v6r), the check resolution process may have been improperly handling cycles when performing exclusion (`but not`). Previously, it was effectively treating all cycles as `allowed=false`, however this isn't correct in all cases. When the right-side of the `but not` operator yields a cycle, we should _not_ treat that as a falsey outcome because it may unintentionally authorize the check request. This PR fixes this issue by propagating the detection of cycles up the `expand` call chain and in cases of exclusion, use that detected cycle to prevent unintentional authorization.

## References
Related CVE: https://github.com/openfga/openfga/security/advisories/GHSA-8cph-m685-6v6r

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/server/commands/listusers/list_users_rpc_test.go [hunk 1]
```diff
@@ -2252,8 +2252,8 @@ func TestListUsersCycleDetection(t *testing.T) {
 				},
 				visitedUsersetsMap: visitedUsersets,
 			}, channelWithResults)
-			if err != nil {
-				channelWithError <- err
+			if resp.err != nil {
+				channelWithError <- resp.err
 				return
 			}
 			channelDone <- struct{}{}
```
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc_test.go:2259]: Can we add this test as well please? It makes sure that if we experience cycles through independent object_ids and with a non-stratifiable resolution on the `document:1#viewer` and `document:2#blocked` relationship then we still yield the residual relationships for `user:jon`.

```go
		{
			name: "exclusion_with_chained_negation",
			req: &openfgav1.ListUsersRequest{
				Object:   &openfgav1.Object{Type: "document", Id: "2"},
				Relation: "viewer",
				UserFilters: []*openfgav1.UserTypeFilter{
					{
						Type: "user",
					},
				},
			},
			model: `model
			  schema 1.1

			type user

			type document
			  relations
			    define unblocked: [user]
				define blocked: [user, document#viewer] but not unblocked
				define viewer: [user, document#blocked] but not blocked
			`,
			tuples: []*openfgav1.TupleKey{
				tuple.NewTupleKey("document:1", "viewer", "document:2#blocked"),
				tuple.NewTupleKey("document:2", "blocked", "document:1#viewer"),
				tuple.NewTupleKey("document:2", "viewer", "user:jon"),
				tuple.NewTupleKey("document:2", "unblocked", "user:jon"),
			},
			expectedUsers: []string{"user:jon"},
		},
```

File: pkg/server/commands/listusers/list_users_rpc_test.go [hunk 2]
```diff
@@ -1507,7 +1507,7 @@ func TestListUsersExclusion(t *testing.T) {
 				tuple.NewTupleKey("group:1", "blocked", "group:1#member"),
```
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc_test.go:1507]: Can we add this test as well please? It makes sure that if we experience cycles through independent object_ids and with a non-stratifiable resolution on the `document:1#viewer` and `document:2#blocked` relationship then we still yield the residual relationships for `user:jon`.

```go
		{
			name: "exclusion_with_chained_negation",
			req: &openfgav1.ListUsersRequest{
				Object:   &openfgav1.Object{Type: "document", Id: "2"},
				Relation: "viewer",
				UserFilters: []*openfgav1.UserTypeFilter{
					{
						Type: "user",
					},
				},
			},
			model: `model
			  schema 1.1

			type user

			type document
			  relations
			    define unblocked: [user]
				define blocked: [user, document#viewer] but not unblocked
				define viewer: [user, document#blocked] but not blocked
			`,
			tuples: []*openfgav1.TupleKey{
				tuple.NewTupleKey("document:1", "viewer", "document:2#blocked"),
				tuple.NewTupleKey("document:2", "blocked", "document:1#viewer"),
				tuple.NewTupleKey("document:2", "viewer", "user:jon"),
				tuple.NewTupleKey("document:2", "unblocked", "user:jon"),
			},
			expectedUsers: []string{"user:jon"},
		},
```

Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1507]: Please add this test as well, and it is currently failing. This is a tricky corner case involving non-stratifiable queries through exclusion.

```
{
	name: "non_stratifiable_exclusion_containing_cycle",
	req: &openfgav1.ListUsersRequest{
		Object:   &openfgav1.Object{Type: "document", Id: "1"},
		Relation: "viewer",
		UserFilters: []*openfgav1.UserTypeFilter{
			{
				Type:     "document",
				Relation: "blocked",
			},
		},
	},
	model: `model
		schema 1.1

	type user

	type document
		relations
		define blocked: [user, document#viewer]
		define viewer: [user, document#blocked] but not blocked
	`,
	tuples: []*openfgav1.TupleKey{
		tuple.NewTupleKey("document:1", "viewer", "document:2#blocked"),
		tuple.NewTupleKey("document:2", "blocked", "document:1#viewer"),
	},
	expectedUsers: []string{},
},
```

For the same `Check(document:1#viewer@document:2#blocked)` this should return `{allowed: false}` as it is a non-stratifable query.

Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1507]: This is a good point. The missing piece was handling the base-cycle case. There are zero situations of negation where a cycle can return a result in ListUsers.

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 3]
```diff
@@ -549,6 +573,12 @@ func (l *listUsersQuery) expandExclusion(
 		subtractFoundUsersMap[key] = struct{}{}
 	}
 
+	if baseHasCycle || subtractHasCycle {
+		return expandResponse{
+			err: nil,
+		}
+	}
```
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:580]: Can you explain in code comments why this is not

```go
if baseHasCycle || subtractHasCycle {
		return expandResponse{
		        hasCycle: true,
			err: nil,
		}
	}
```

?

Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go:580]: Because exclusion contains the only bespoke treatment of cycle, everywhere else we consider it a falsey outcome. Once we make a determination within the exclusion handler, we're able to properly handle the case and do not need to propagate the existence of a cycle to an upstream handler.

Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc.go:580]: > in code comments

😄 so we don't forget

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 4]
```diff
@@ -415,24 +431,30 @@ LoopOnIterator:
 			rewrittenReq := req.clone()
 			rewrittenReq.Object = &openfgav1.Object{Type: userObjectType, Id: userObjectID}
 			rewrittenReq.Relation = userRelation
-			return l.expand(ctx, rewrittenReq, foundUsersChan)
+			resp := l.expand(ctx, rewrittenReq, foundUsersChan)
+			if resp.hasCycle {
+				hasCycle = true
+			}
```
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:437]: nit: can simplify `hasCycle = hasCycle || resp.hasCycle`

Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go:437]: Interesting enough, that causes a data race.

Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc.go:437]: Oh! You have the data race even without my suggestion. `var hasCycle bool` should be an `atomic.Bool`
